distribution/pull_v2_unix.go

```diff
@@ -17,3 +29,69 @@ func (ld *v2LayerDescriptor) open(ctx context.Context) (distribution.ReadSeekClo
 	blobs := ld.repo.Blobs(ctx)
 	return blobs.Open(ctx, ld.digest)
 }
+
+func configurePolicyContext() (*signature.PolicyContext, error) {
+	defaultPolicy, err := signature.DefaultPolicy(nil)
+	if err != nil {
+		return nil, err
+	}
+	pc, err := signature.NewPolicyContext(defaultPolicy)
+	if err != nil {
+		return nil, err
+	}
+	return pc, nil
+}
+
+func (p *v2Puller) checkTrusted(c gctx.Context, ref reference.Named) (reference.Named, error) {
+	p.originalRef = ref
+	// we can't use upstream docker/docker/reference since in projectatomic/docker
+	// we modified docker/docker/reference and it's not doing any normalization.
+	// we instead forked docker/docker/reference in containers/image and we need
+	// this parsing here to make sure signature naming checks are consistent.
+	dockerRef, err := containersImageRef.ParseNamed(ref.String())
+	if err != nil {
+		return nil, err
+	}
+	imgRef, err := docker.NewReference(dockerRef)
+	if err != nil {
+		return nil, err
+	}
+	isSecure := (p.endpoint.TLSConfig == nil || !p.endpoint.TLSConfig.InsecureSkipVerify)
+	authConfig := registry.ResolveAuthConfig(p.config.AuthConfigs, p.repoInfo.Index)
+	dockerAuthConfig := types.DockerAuthConfig{
+		Username: authConfig.Username,
+		Password: authConfig.Password,
+	}
+	ctx := &types.SystemContext{
+		DockerInsecureSkipTLSVerify: !isSecure,
+		DockerAuthConfig:            &dockerAuthConfig,
+		DockerRegistryUserAgent:     dockerversion.DockerUserAgent(c),
+	}
+	img, err := imgRef.NewImage(ctx)
```

Comment:
This could now be more secure with a `imgRef.NewImageSource`→`image.UnparsedFromSource`, the way `copy.Image` does it; it would skip all of the risky manifest parsing `image.FromUnparsedImage` is doing which nothing here seems to need.  (AFAICT this change wouldn’t affect the behavior with manifest lists, but that would be worth checking).